### PR TITLE
[Docs] General extensibility and implementation of proactive strategies

### PR DESCRIPTION
**Pull request by martintmk**
### Details on the issue fix or feature implementation

Contributes to #1091 

### Confirm the following

- [x]  I started this PR by branching from the head of the default branch
- [x]  I have targeted the PR to merge into the default branch
- [x]  I have included unit tests for the issue/feature
- [x]  I have successfully run a local build

### CODE REVIEW

File: docs/extensibility/index.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+<!-- endSnippet -->
+
+## Arguments
+
+Arguments are used by individual delegate types to flow information to the consumer. Arguments should always have an `Arguments` suffix and include a `Context` property. Using arguments boosts the extensibility and maintainability of the API, as adding new members becomes a non-breaking change.  For proactive strategies, the arguments structure might resemble:
+
+<!-- snippet: ext-proactive-args -->
+```cs
+// Arguments-based structs encapsulate information about particular event that occurred inside resilience strategy.
+// They cna expose any properties that are relevant to the event.
```
Comment (adamnova) [inline docs/extensibility/index.md:95]:
```suggestion
// They can expose any properties that are relevant to the event.
```

File: docs/extensibility/proactive-strategy.md [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+```
+<!-- endSnippet -->
+
+Arguments should always have an `Arguments` suffix and include a `Context` property. Using arguments boosts the extensibility and maintainability of the API, as adding new members becomes a non-breaking change. The `ThresholdExceededArguments` provides details about the actual execution time and threshold, allowing listeners to respond to this event or supply a custom callback for such situations.
+
+## Options
+
+In the previous section, we implemented the `TimingResilienceStrategy`. Now, it's time to integrate it with Polly and its public API.
+
+Let's define the public `TimingResilienceStrategy` to configure our strategy:
```
Comment (adamnova) [inline docs/extensibility/proactive-strategy.md:97]:
```suggestion
Let's define the public `TimingStrategyOptions` to configure our strategy:
```

File: docs/extensibility/proactive-strategy.md [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+<!-- snippet: ext-proactive-options -->
+```cs
+public class TimingStrategyOptions : ResilienceStrategyOptions
+{
+    public TimingStrategyOptions()
+    {
+        // It's recommended to set the default name for the options so
+        // the consumer can get additional information in the telemetry.
```
Comment (adamnova) [inline docs/extensibility/proactive-strategy.md:106]: What additional information? If it's just the name I would probably say "the consumer can see it in the telemetry" if it allows access to more things, an example could be handy to help build a picture.

File: docs/extensibility/proactive-strategy.md [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+```
+<!-- endSnippet -->
+
+Options represent our public contract with the consumer. By using them, we can easily add new members without breaking changes and perform validation consistently.
+
+## Extensions
+
+So far, we have:
+
+- Public `TimingResilienceStrategy` and the public arguments associated with them.
```
Comment (adamnova) [inline docs/extensibility/proactive-strategy.md:130]: `TimingStrategyOptions`?

File: docs/extensibility/index.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+Regardless of whether the strategy is reactive or proactive, every new resilience strategy should include the following components:
+
+- Options detailing the strategy's configuration. These should inherit from [`ResilienceStrategyOptions`](xref:Polly.ResilienceStrategyOptions).
+- Extensions for `ResiliencePipelineBuilder` or `ResiliencePipelineBuilder<T>`.
+- Custom argument types for delegates that contain information about a specific event.
+
+The strategy options contain properties of following types:
+
+- **Primitive types**: Such as `int`, `bool`, `TimeSpan`, etc.
```
[the remark this answers is not in the document]
Comment (martintmk) [reply docs/extensibility/index.md:20]: I'll use "common" types here.

[the remark this answers is not in the document]
Comment (adamnova) [reply docs/extensibility/index.md:20]: Sure that works, just to give the reader some idea what could be behind 'etc.'

File: docs/extensibility/proactive-strategy.md (hunk 3 above)
Comment (martintmk) [reply docs/extensibility/proactive-strategy.md]: Reworded, as these are not just for telemetry. They can be directly consumed.

**Review by adamnova — APPROVED** [review]

**Pull request merged**